File: insight-lang/src/main/java/com/github/lonelylockley/archinsight/lexer/LexerState.java
package com.github.lonelylockley.archinsight.lexer;

import java.util.Objects;

public class LexerState {

    private boolean wasText = false;
    private int indentation = 0;

    public LexerState() {
    }

    protected LexerState(boolean wasText, int indentation) {
        this.wasText = wasText;
        this.indentation = indentation;
    }

    public boolean wasText() {
        return wasText;
    }

    public void setWasText() {
        this.wasText = true;
    }

    public void resetWasText() {
        this.wasText = false;
    }

    public int getIndentation() {
        return indentation;
    }

    public void incIndentation() {
        this.indentation++;
    }

    public void decIndentation() {
        this.indentation--;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        LexerState that = (LexerState) o;
        return wasText == that.wasText && indentation == that.indentation;
    }

    @Override
    public int hashCode() {
        return Objects.hash(wasText, indentation);
    }

    @Override
    public Object clone() {
        return new LexerState(this.wasText, this.indentation);
    }

}

File: editor-ui/frontend/src/LexerState.ts
import { Token } from 'antlr4ng';
import { languages } from 'monaco-editor';
import IState = languages.IState;

class LexerState implements IState {

    private _wasText: boolean = false;
    private indentation: number = 0;

     constructor() {
     }

     public wasText(): boolean {
         return this._wasText;
     }

     public setWasText() {
         this._wasText = true;
     }

     public resetWasText() {
         this._wasText = false;
     }

     public getIndentation(): number {
         return this.indentation;
     }

     public incIndentation() {
         this.indentation++;
     }

     public decIndentation() {
         this.indentation--;
     }

     public clone(): IState {
         var res = new LexerState();
         res._wasText = this._wasText;
         res.indentation = this.indentation;
         return res;
     }

     equals(that: IState): boolean {
         if (that instanceof LexerState) {
             return this._wasText === that._wasText && this.indentation === that.indentation;
         }
         else {
             return false;
         }
     }

}

export default LexerState;